### ISSUE TITLE
Potential confusion for end users and implementers: host language conflicts handled inconsistently

### ISSUE THREAD

**Issue by karlgroves** [w3c/aria#486]
For reference, I'll be discussing the following: https://www.w3.org/TR/wai-aria-1.1/#host_general_conflict. Particularly the following:

> When a host language declares a WAI-ARIA attribute to be in direct semantic conflict with a native attribute for a given element, user agents MUST ignore the WAI-ARIA attribute and instead use the host language attribute with the same implicit semantic.

I agree completely that the host language should take precedence. The problem is that ARIA contradicts itself in the case of `role` and `aria-label`/ `aria-labelledby`.

Consider the following:
```<input type="text" required aria-required="false">```

In the above case, the HTML `required` attribute wins as it should.

However, in the following case, the host language does not win. It is identified as a radio button
```
<input type="text" role="radio">
```

The same goes for the following case where a good explicit label is overridden by `aria-label`:

```
<label for="fname">First Name</label>
<input type="text" name="fname" id="fname" aria-label="Last Name">
```

Edit: Removed last sentence from the original entry. Admittedly poor form.

**Comment by richschwer** [comment]
First Karl, HTML host language MUST state that you cannot use the attribute . It MUST be prohibited. This is where host conflicts are determined. aria-required is prohibited in the case you mention as there is no additional value to adding, but it does create a real conflict. The ARIA label attributes are global and having special cases for all the instances where you can use it or not is crazy - especially when you think about name computation from content. We have said all along that aria-label and aria-labelledby *win* as the author has explicitly chosen to override the normal labeling mechanism to provide the accessible name. This is not a new development in ARIA 1.1 and you should know that. There may be instances why the author has chosen to provide a different accessible name that that provided by <label for=""> such as the need to concatenate additional text to the label in the formation of the accessible name to help someone who is blind. 

If the HTML working group were to restrict the use of aria labeling mechanisms I would be the first to formally object to it. Since you are an accessibility consultant my advice to you is to properly educate developers when they should be overriding the label element usage. If you like we can have the ARIA authoring practices people provide additional guidance here. 

Finally, Karl, you don't need to educate the ARIA working group about how hard it is to get authors to care about accessibility. We are sympathetic. Many of us are responsible for driving accessibility in large companies. IBM alone had around 400,000 employees. You might consider reserving your last comment for other working groups. :-)

**Comment by karlgroves** [comment]
Rich,  

Thanks for the response.

> This is not a new development in ARIA 1.1 and you should know that. 

Of course. However, since 1.1 is the current work-in-progress it appears most appropriate to comment here.

> The ARIA label attributes are global and having special cases for all the instances where you can use it or not is crazy - especially when you think about name computation from content.

I'm sorry, but I don't see what's "crazy" about having the host language's mechanism for labelling take precedence over ARIA. In fact, the special case in this instance is this specific behavior of having ARIA override the host language. 

To reiterate, the spec itself states: 
> When a host language declares a WAI-ARIA attribute to be in direct semantic conflict with a native attribute for a given element, user agents MUST ignore the WAI-ARIA attribute and instead use the host language attribute with the same implicit semantic.

So, the "special case" here is the behavior of the ARIA labelling technique overriding the host language when elsewhere it is stated that it shouldn't.

> There may be instances why the author has chosen to provide a different accessible name that that provided by such as the need to concatenate additional text to the label in the formation of the accessible name to help someone who is blind.

Of course. I've had to do this myself many times. Your above statement isn't incompatible with my argument. ARIA provides mechanisms for creating such a label when the host language's existing semantics don't meet the author's requirements. The same goes for `role`.

In short, it seems to make the most sense to have the host language win across the board.

**Comment by stevefaulkner** [comment]
>In short, it seems to make the most sense to have the host language win across the board.

This would mean that across the board roles would not work, as even `<div>`s have [a role](https://w3c.github.io/html-aam/#div-element) exposed in accessibility APIs, it is just a role that is not conveyed to users by AT.

It would have been useful to have this discussion when ARIA was first being defined and implemented, it does not make a lot of sense now as changes such as this would not be backwards compatible and would break many implementations.

**Comment by karlgroves** [comment]
> This would mean that across the board roles would not work, as even <div>s have a role exposed in accessibility APIs, it is just a role that is not conveyed to users by AT.

True. I should've been more clear that I meant specific roles necessary for controls. Using the example I provided in the original post, for instance:  ```<input type="text" role="radio">```  this should not be possible.

> It would have been useful to have this discussion when ARIA was first being defined and implemented, it does not make a lot of sense now as changes such as this would not be backwards compatible and would break many implementations.

Implementers change things quite frequently. I don't want to discount the difficulty of doing such a thing, I don't see this as a blocker to the change.  Was it not long ago that implementers had to be asked not to expose `<section>` as landmarks?

**Comment by richschwer** [comment]
Karl, this is the working group response to your comment:  In general the ARIA working group believe that if the author has gone to such great lenghts to repurpose a host language element such that they would require an ARIA role to be applied to make it accessible they should be able to. That said, we do have a provision in the ARIA spec. to address conflicts with host languanges in this section of ARIA: http://rawgit.com/w3c/aria/master/aria/aria.html#host_general_conflict . When a WAI-ARIA role is provided, user agents MUST use the semantic of the WAI-ARIA role for processing, not the native semantic, unless the role requires WAI-ARIA states and properties whose attributes are explicitly forbidden on the native element by the host language. Values for roles do not conflict in the same way as values for states and properties (for example, the HTML 'checked' attribute and the 'aria-checked' attribute could have conflicting values), and authors are expected to have valid reason to provide a WAI-ARIA role even on elements that would not normally be repurposed.

**Comment by stevefaulkner** [comment]
> True. I should've been more clear that I meant specific roles necessary for controls. Using the example I provided in the original post, for instance: <input type="text" role="radio"> this should not be possible.

So there is a BIG difference between saying 'all roles' to saying in _special cases_, which is what I thought you were remonstrating about in your original comment.

There are a number of ways in which HTML defines native semantics that must not be overriden by ARIA. By defining conforming uses of ARIA in HTML: http://w3c.github.io/html-aria/ and by defining implementation requirements https://w3c.github.io/html-aam/

In the case of:
`<input type="text" role="radio"> `

It is non conforming, but is allowed in implementations. The disparity here is due to it being more practical to prohibit for authors than asking implementers to special case it. A route towards defining such an implementation requirement would be if it was found that dev's were writing this code and it was causing issues for users, I am unaware of this occuring, but would happily take it up if it was (implementation isssues should be filed against https://w3c.github.io/html-aam/).

> Implementers change things quite frequently. I don't want to discount the difficulty of doing such a thing, I don't see this as a blocker to the change. Was it not long ago that implementers had to be asked not to expose <section> as landmarks?

Implementers do not generally implement **breaking changes** unless they are really serious (security issues come to mind). In the case you mention, there was no actual change, `<section>` was and is exposed with a `region` role, but it is [strongly suggested that AT implementers](https://w3c.github.io/html-aam/#el-section) do not convey the presence of a `section` as a `region` unless the `section` has an accessible name. This suggestion was introduced (not a long time after the implict role for section was implemented) after feedback from SR users saying that the announcing of `section` was damned annoying as developers were using it like a `div`.
